cmd/sink/event_data.go

```diff
@@ -0,0 +1,138 @@
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/cloudevents/sdk-go/v2/types"
+)
+
+const (
+	apiVersionExtension = "apiversion"
+	kindExtension       = "kind"
+	nameExtension       = "name"
+	namespaceExtension  = "namespace"
+
+	writeQuery = "INSERT INTO ArchiveMeta (api_version, kind, name, namespace, resource_version, data, created_ts, updated_ts) Values ($1, $2, $3, $4, $5, $6, $7, $8)"
+)
+
+// Represents the fields in a resource's metadata that are need to write a resource into the database
+type ResourceMetadata struct {
+	ResourceVersion string `json:"resourceVersion"`
+}
+
+// Respresents the fields in the data field of a cloudevent to write a resource into the database
+type ResourceData struct {
+	Created     string           `json:"firstTimestamp"`
+	LastUpdated string           `json:"lastTimestamp"`
+	Metadata    ResourceMetadata `json:"metadata"`
+}
+
+// A container for all of the data required for writing a Kubernetes resource into the database
+type ArchiveEntry struct {
+	Event      cloudevents.Event
+	Data       ResourceData
+	ApiVersion string
+	Kind       string
+	Name       string
+	Namespace  string
+}
+
+// implement the Stringers interface so that golang can convert ArchiveEntries to string
+func (entry *ArchiveEntry) String() string {
+	return string(entry.Event.Data())
+}
+
+// the timestamp when the resource was created
+func (entry *ArchiveEntry) Created() string {
+	return entry.Data.Created
+}
+
+// the timestamp when the resource was last updated
+func (entry *ArchiveEntry) LastUpdated() string {
+	return entry.Data.LastUpdated
+}
+
+// the version of the resources
+func (entry *ArchiveEntry) ResourceVersion() string {
+	return entry.Data.Metadata.ResourceVersion
+}
+
+// checks that the cloudevents has the appropriate extensions set and have values that are the right type. Additionally
+// it checks that the cloudevent's data has the necessary fields. If all conditions are not met, it returns an error
+func NewArchiveEntry(event cloudevents.Event) (*ArchiveEntry, error) {
+	eventExtensions := event.Extensions()
+	apiVersion, err := types.ToString(eventExtensions[apiVersionExtension])
+	if err != nil {
+		return nil, err
+	}
+	kind, err := types.ToString(eventExtensions[kindExtension])
+	if err != nil {
+		return nil, err
+	}
+	name, err := types.ToString(eventExtensions[nameExtension])
+	if err != nil {
+		return nil, err
+	}
+	namespace, err := types.ToString(eventExtensions[namespaceExtension])
+	if err != nil {
+		return nil, err
+	}
+	payload := ResourceData{}
+	err = event.DataAs(&payload)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ArchiveEntry{
+			Event:      event,
+			Data:       payload,
+			ApiVersion: apiVersion,
+			Kind:       kind,
+			Name:       name,
+			Namespace:  namespace,
+		},
+		nil
+}
+
+// Writes the kubernetes resource held by entry to the database using the provided connection string. Timeouts and
```

Comment:
The method takes a connection (dbConn), not a connection string:

```suggestion
// Writes the kubernetes resource held by entry to the database using the provided DB connection. Timeouts and
```

Reply:
fixed